File: data/models/order_model.dart
class OrderItemModel {
  final String productId;
  final String productName;
  final int priceAtPurchase;
  final int quantity;
  final String imageUrl;

  OrderItemModel({
    required this.productId,
    required this.productName,
    required this.priceAtPurchase,
    required this.quantity,
    required this.imageUrl,
  });

  factory OrderItemModel.fromJson(Map<String, dynamic> json) {
    return OrderItemModel(
      productId: json['productId'] ?? '',
      productName: json['productName'] ?? '',
      priceAtPurchase: json['priceAtPurchase'] ?? 0,
      quantity: json['quantity'] ?? 1,
      imageUrl: json['imageUrl'] ?? '',
    );
  }

  int get subtotal => priceAtPurchase * quantity;
}

class OrderModel {
  final String id;
  final String orderId;
  final String userId;
  final List<OrderItemModel> items;
  final int totalAmount;
  final String status;
  final String shippingAddress;
  final String? paymentToken;
  final String? paymentUrl;
  final DateTime createdAt;

  OrderModel({
    required this.id,
    required this.orderId,
    required this.userId,
    required this.items,
    required this.totalAmount,
    required this.status,
    required this.shippingAddress,
    this.paymentToken,
    this.paymentUrl,
    required this.createdAt,
  });

  factory OrderModel.fromJson(Map<String, dynamic> json) {
    final itemsList = json['items'] as List? ?? [];
    return OrderModel(
      id: json['_id'] ?? '',
      orderId: json['orderId'] ?? '',
      userId: json['userId'] ?? '',
      items: itemsList.map((item) => OrderItemModel.fromJson(item)).toList(),
      totalAmount: json['totalAmount'] ?? 0,
      status: json['status'] ?? 'Pending',
      shippingAddress: json['shippingAddress'] ?? '',
      paymentToken: json['paymentToken'],
      paymentUrl: json['paymentUrl'],
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
    );
  }

  bool get isPending => status == 'Pending';
  bool get isPaid => status == 'Paid';
  bool get isShipped => status == 'Shipped';
  bool get isCompleted => status == 'Completed';
  bool get isCancelled => status == 'Cancelled';
  
  bool get canBeCancelled => isPending;
}

File: src/models/Order.model.js
const mongoose = require('mongoose');
const { ORDER_STATUS } = require('../config/constants');

const OrderItemSchema = new mongoose.Schema({
  productId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Product',
    required: true
  },
  productName: {
    type: String,
    required: true
  },
  priceAtPurchase: {
    type: Number,
    required: true,
    min: 0
  },
  quantity: {
    type: Number,
    required: true,
    min: 1
  },
  imageUrl: {
    type: String
  }
}, { _id: false });

const OrderSchema = new mongoose.Schema(
  {
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true
    },
    orderId: {
      type: String,
      required: true,
      unique: true
    },
    items: [OrderItemSchema],
    subtotal: { 
      type: Number, 
      required: true, 
      default: 0 
    },
    shippingFee: { 
      type: Number, 
      required: true, 
      default: 0 
    },
    totalAmount: {
      type: Number,
      required: true,
      min: 0
    },
    status: {
      type: String,
      enum: Object.values(ORDER_STATUS),
      default: ORDER_STATUS.PENDING
    },
    shippingAddress: {
      type: String,
      required: true
    },
    paymentToken: {
      type: String
    },
    paymentUrl: {
      type: String
    }
  },
  { 
    timestamps: true,
    collection: 'orders'
  }
);

OrderSchema.index({ userId: 1, createdAt: -1 });
OrderSchema.index({ status: 1 });

module.exports = mongoose.model('Order', OrderSchema);